Mathlib/Algebra/Ring/Defs.lean

```diff
@@ -0,0 +1,466 @@
+-/
+
+
+universe u v w x
+
+variable {α : Type u} {β : Type v} {γ : Type w} {R : Type x}
+
+open Function
+
+/-!
+### `distrib` class
+-/
+
+
+/-- A typeclass stating that multiplication is left and right distributive
+over addition. -/
+class Distrib (R : Type _) extends Mul R, Add R where
+  protected left_distrib : ∀ a b c : R, a * (b + c) = a * b + a * c
+  protected right_distrib : ∀ a b c : R, (a + b) * c = a * c + b * c
+#align distrib Distrib
+
+/-- A typeclass stating that multiplication is left distributive over addition. -/
+class LeftDistribClass (R : Type _) [Mul R] [Add R] where
+  protected left_distrib : ∀ a b c : R, a * (b + c) = a * b + a * c
+#align left_distrib_class LeftDistribClass
+
+/-- A typeclass stating that multiplication is right distributive over addition. -/
+class RightDistribClass (R : Type _) [Mul R] [Add R] where
+  protected right_distrib : ∀ a b c : R, (a + b) * c = a * c + b * c
+#align right_distrib_class RightDistribClass
+
+-- see Note [lower instance priority]
+instance (priority := 100) Distrib.leftDistribClass (R : Type _) [Distrib R] : LeftDistribClass R :=
+  ⟨Distrib.left_distrib⟩
+#align distrib.left_distrib_class Distrib.leftDistribClass
+
+-- see Note [lower instance priority]
+instance (priority := 100) Distrib.rightDistribClass (R : Type _) [Distrib R] :
+    RightDistribClass R :=
+  ⟨Distrib.right_distrib⟩
+#align distrib.right_distrib_class Distrib.rightDistribClass
+
+theorem left_distrib [Mul R] [Add R] [LeftDistribClass R] (a b c : R) :
+    a * (b + c) = a * b + a * c :=
+  LeftDistribClass.left_distrib a b c
+#align left_distrib left_distrib
+
+alias left_distrib ← mul_add
+
+theorem right_distrib [Mul R] [Add R] [RightDistribClass R] (a b c : R) :
+    (a + b) * c = a * c + b * c :=
+  RightDistribClass.right_distrib a b c
+#align right_distrib right_distrib
+
+alias right_distrib ← add_mul
+
+theorem distrib_three_right [Mul R] [Add R] [RightDistribClass R] (a b c d : R) :
+    (a + b + c) * d = a * d + b * d + c * d := by simp [right_distrib]
+#align distrib_three_right distrib_three_right
+
+/-!
+### Semirings
+-/
+
+
+/-- A not-necessarily-unital, not-necessarily-associative semiring. -/
+class NonUnitalNonAssocSemiring (α : Type u) extends AddCommMonoid α, Distrib α, MulZeroClass α
+#align non_unital_non_assoc_semiring NonUnitalNonAssocSemiring
+
+/-- An associative but not-necessarily unital semiring. -/
+class NonUnitalSemiring (α : Type u) extends NonUnitalNonAssocSemiring α, SemigroupWithZero α
+#align non_unital_semiring NonUnitalSemiring
+
+/-- A unital but not-necessarily-associative semiring. -/
+class NonAssocSemiring (α : Type u) extends NonUnitalNonAssocSemiring α, MulZeroOneClass α,
+    AddCommMonoidWithOne α
+#align non_assoc_semiring NonAssocSemiring
+
+class Semiring (α : Type u) extends NonUnitalSemiring α, NonAssocSemiring α, MonoidWithZero α
+#align semiring Semiring
+
+-- Porting note: no OfNat α 2
+/-
```

Comment:
Please leave things like this uncommented. For things that still need to be ported, error messages in the build are more helpful than "Porting note", which should be intended for notes that will be committed as part of the PR.